날씨 정보 [Step2] 잼킹, 호싱잉

Hosinging opened the pull request:
안녕하세요. 라자냐 @wonhee009 

# 시뮬레이터상 설정된 좌표로 현재 주소 확인 및 날씨 데이터 자료 가져오기 확인
<img width="1130" alt="스크린샷 2021-10-05 오후 8 29 28" src="https://user-images.githubusercontent.com/70311145/136017028-3523cdd0-beec-4c45-bc6d-0ef4b2fcded1.png">

# 실물 디바이스에서 설정된 좌표로 현재 주소 확인 및 날씨 데이터 자료 가져오기 확인
<img width="730" alt="스크린샷 2021-10-05 오후 8 31 23" src="https://user-images.githubusercontent.com/70311145/136017040-885ae145-d2ec-42de-8223-24b10365ac02.png">

# 궁금했던 점 & 고민했던 점

> URI

uri를 생성해주는 부분을 아래와 같이 enum안에 계산 속성, 그리고 URI를 만들어주는 메서드를 통해 생성해주는 방법과
그 아래에 있는 것과 같이 enum안에 타입 속성을 통해 선언해주고, 이를 메서드에서 활용하는 방법 둘 중에서 고민을 해봤습니다.
둘다 문자열을 다루는 부분이기 때문에 문자열을 기본값으로 갖는 enum을 활용해볼까도 고민을 같이 해봤습니다.
enum안에서 타입 속성을 사용하면 '대소문자 없는 열거형을 사용하면 실수로 인스턴스화할 수 없고 순수 네임스페이스로 작동한다.'라는 글을 보기도 했고, 첫번째 방식이나 다른 방식으로 했을 경우 일을 위한 일을 만드는 것 같은 느낌이들어 최종적으로는 두번째 코드로 URI생성 부분을 정리해봤는데요. 만드는 방식이 다양하기 떄문에 만들기 나름이라는 생각이 들긴하지만, 그래도 적절한 방법이었는지 궁금하고 라자냐라면 이런 코드는 어떤 방식으로 처리하실지 궁금합니다.

```swift
enum URI {
    case url
    case currentPath
    case fiveDayPath
    case appId
    case unit
    case lang
    
    var element: String {
        switch self{
        case .url:
            return "https://api.openweathermap.org/"
        case .currentPath:
            return "data/2.5/weather?"
        case .fiveDayPath:
            return "data/2.5/forecast?"
        case .appId:
            return "&appid=af361cc4ac7bf412119174d64ba296ff"
        case .unit:
            return "&units=metric"
        case .lang:
            return "&lang=kr"
        }
    }
}

func generateURI(t: [URI]) -> String {
    var uri = ""
    for item in t {
        uri += item.element
    }
    return uri
}

let uri = generateURI(t: [.url, .currentPath, .appId, .unit, .lang])
```

```swift
enum URI {
    static let host = "https://api.openweathermap.org/"
    static let currentPath = "data/2.5/weather?"
    static let fiveDaysPath = "data/2.5/forecast?"
    static let appID = "&appid=af361cc4ac7bf412119174d64ba296ff"
    static let units = "&units=metric"
    static let lang = "&lang=kr"
}

 private func generateURI(path: Bool, location: CLLocation) -> String {
    let latString = "lat=\(location.coordinate.latitude)"
    let lonString = "&lon=\(location.coordinate.longitude)"

    isPathCurrent = path

    var uri: String {
        if isPathCurrent {
            return "\(URI.host)\(URI.currentPath)\(latString)\(lonString)\(URI.appID)\(URI.units)\(URI.lang)"
        } else {
            return "\(URI.host)\(URI.fiveDaysPath)\(latString)\(lonString)\(URI.appID)\(URI.units)\(URI.lang)"
        }
    }
    return uri
}
```


> 위치정보를 다루는 부분

코드를 작성하다보니 WeatherDataManager와 WeatherListViewController 두 부분에서 위치 정보를 다뤄야하는 상황이 이뤄졌었는데요.
두 부분에서 위치 정보를 다루는 것이 적절한가에 대한 고민이 있었습니다. 현재 구조로는 두 부분에서 다뤄야하는 형태라고 생각이되는데 개선을 해본다면 어떤 부분을 고려해보면 좋을지 궁금합니다.

jaemuYeo committed:
chor: ViewController -> WeatherListViewController 네이밍 변경

Hosinging committed:
feat: CLLocationManager인스턴스 생성 및 델리게이트 지정 메서드 구현

- info.plist 파일 내 Location When in Use Usage Description추가

Hosinging committed:
feat: 현재날씨, 5일 날씨를 가져오는 메서드 리팩터링

- URI타입 추가 및 관련 코드 수정
- generateURI(path:location)메서드 구현 및 호출 코드 추가

Hosinging committed:
feat: locationManager를 초기화, 현재 좌표 받아오기, Autorination status에 따른 분기 처리 코드 추가

- locationManager초기화하는 generateLocationManager()추가
- 현재 좌표 받아오는 generateCurrentCoordinate()추가
- 권한에 따라 분기 처리해주는 locationManager(_:didChangeAuthorization)추가

Hosinging committed:
chore: 타입 내 메서드 익스텐션으로 분리, 접근제어자 추가

jaemuYeo committed:
feat: 위치 정보를 가져와서 주소로 변경하는 코드 구현

- 좌표를 가져오는 bringCoordinates() 메서드 구현
- 좌표를 주소로 변환해주는 converToAddress(with:)메서드 구현
- viewDidLoad() 메서드에 필요한 메서드 호출

wonhee009 commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -30,14 +31,33 @@ enum APIError: LocalizedError {
+    static let host = "https://api.openweathermap.org/"
+    static let currentPath = "data/2.5/weather?"
+    static let fiveDaysPath = "data/2.5/forecast?"
+    static let appID = "&appid=af361cc4ac7bf412119174d64ba296ff"
+    static let units = "&units=metric"
+    static let lang = "&lang=kr"
```
`appID`, `units`, `lang`는 파라미터들인거 같은데요.
파라미터가 추가된다면 `enum URI`에 계속 추가될텐데 가독성 측면에서 떨어질거 같습니다. 또한, 파라미터들이 늘어남에 따라 구현하면서 에러가 생길 가능성도 높을거 같습니다.
파라미터들을 따로 분리하는건 어떤가요?

'=00'으로 뒤에 파라미터값이 들어가고 있는데 현재는 `lang=kr`로 받아오고 있지만 기능이 추가되어 `lang=en`으로 받아와야 하는 경우에는 현재 구조에서는 타입 프로퍼티가 더 추가되어야할 것 같은데요.
파라미터를 따로 분리하고 파라미터 값을 외부에서 주입하는 방식은 어떨까요?

Hosinging replied:
'기능이 추가되는 경우를 고려해서 파라미터 값을 외부에서 주입해주는 방식'을 반영해보기 위해 URLBuilder타입을 구현하고, build메서드 내에서 URLResource타입을 받아서 urlcomponents의 schem과 host에 바인딩을 해주고, queryItems는 URLBuilder타입 내에 있는 queries배열을 통해 URLQueryItem이 유연하게 채워질 수 있게 시도해봤습니다.

'파라미터들이 늘어날 경우'에 대해서는 QueryItem의 경우가 해당이 될 것 같았는데, URLQueryItem생성자를 통해 외부에서 아래 두 메서드에 QueryItem이 될 수 있는 값(키,밸류)을 문자열로 직접 넣어주는 방식과 상수 바인딩 후 넣어주는 방식, 타입의 프로퍼티를 통해 넣어주는 방식 등 고민을 해봤는데요.

```swift 
builder.addQuery(_:)
builder.addQueries(_:)
```
현재는 직접 문자열로 넣는 것도 가능하고, (그외 나머지 방법들은 크게 차이가 없다고 생각이 되어서...)ParamName, AppID, Units, Lang이라는 enum을 각각 만들고 그 타입 프로퍼티로 QueryItem생성시 필요한 키와 값들을 넣어주는 것도 가능하게 구현해봤습니다.

wonhee009 commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -59,6 +79,22 @@ final class WeatherDataManager {
         }
     }
+    
+    private func generateURI(path: Bool, location: CLLocation) -> String {
```
파라미터 이름이 `path`인데 Bool 타입인게 어색한거 같아요.

jaemuYeo replied:
이 부분도 마찬가지로 열거형으로 케이스를 나누면서 수정되었습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -59,6 +79,22 @@ final class WeatherDataManager {
     }
+    
+    private func generateURI(path: Bool, location: CLLocation) -> String {
+        let latString = "lat=\(location.coordinate.latitude)"
+        let lonString = "&lon=\(location.coordinate.longitude)"
```
해당 문자열도 url을 생성하는데 필요한 파라미터라고 생각되는데요. 여기에 정의한 이유가 있을까요?

jaemuYeo replied:
URLComponents 개념을 늦게 파악해서 선택했던 결과물이였던것 같아요.
파라미터를 분리하는 작업에서 해당 부분을 없앴습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -59,6 +79,22 @@ final class WeatherDataManager {
+        isPathCurrent = path
+
+        var uri: String {
+            if isPathCurrent {
+                return "\(URI.host)\(URI.currentPath)\(latString)\(lonString)\(URI.appID)\(URI.units)\(URI.lang)"
+            } else {
+                return "\(URI.host)\(URI.fiveDaysPath)\(latString)\(lonString)\(URI.appID)\(URI.units)\(URI.lang)"
+            }
```
uri 문자열을 여기서 생성해주는데 url에 변화가 있다면 수정할때 어려울거 같은데 어떻게 생각하시나요?

jaemuYeo replied:
URLComponents를 통해서 URL을 받아오는 작업으로 처리하였습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,87 @@
+import CoreLocation
+
+class WeatherListViewController: UIViewController {
+    private var locationManager: CLLocationManager!
+    private var currentLocation: CLLocationCoordinate2D!
```
강제 언래핑 타입을 사용한 이유가 있을까요?

wonhee009 commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        
+        generateLocationManager()
+        bringCoordinates()
+        converToAddress(with: generateCurrentCoordinate())
```
`convertToAddress`일거 같습니다

wonhee009 commented on WeatherForecast/WeatherForecast/WeatherListViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        locationManager.desiredAccuracy = kCLLocationAccuracyBest
+    }
+    
+    private func generateCurrentCoordinate() -> CLLocation {
+        let manager = WeatherDataManager.shared
+        guard let latitude = manager.latitude, let longitude = manager.longitude else { return CLLocation() }
+        let location = CLLocation(latitude: latitude, longitude: longitude)
+        return location
+    }
+    
+    private func bringCoordinates() {
+        if CLLocationManager.locationServicesEnabled() {
+            let currentCoordinate = locationManager.location?.coordinate
+            
+            locationManager.startUpdatingLocation()
+            guard let lat = currentCoordinate?.latitude, let lon = currentCoordinate?.longitude else {
+                return
+            }
+            WeatherDataManager.shared.latitude = lat
+            WeatherDataManager.shared.longitude = lon
+        }
+    }
```
`bringCoordinates()`에서 WeatherDataManager에 위경도를 지정해주고 있는데 `generateCurrentCoordinate()`가 따로 필요한가요?

wonhee009 reviewed (changes requested):
잼킹, 호싱잉 스텝2 수고하셨습니다!🎉

리뷰 남겨두었고, 잼킹과 호싱잉이 남겨둔 질문은 코멘트를 보면 좀 해소될 수 있는 부분들이 있어보여서 수정 이후에 다시 얘기해보면 좋겠습니다🙏

jaemuYeo committed:
chore: 파일 그룹별 정리

jaemuYeo committed:
feat: 쿼리 파라미터 타입별 분리

- APIResource.swift 파일 생성
- ParamName, AppID, Units, Lang 타입 생성 및 상수 저장

Hosinging committed:
feat: 의존성 주입을 통해 URL을 생성해주는 URLBuilder타입 구현

Hosinging committed:
feat: URL생성에 필요한 URLResource타입 구현

Hosinging committed:
feat: URLResource에 필요한 프로퍼티를 정의한 프로토콜 정의 및 채택

Hosinging committed:
refactor: 추가한 URLBuilder객체를 통해 URL을 생성하는 generateURL(path:) 리팩터링

- WeatherDataManager의 Extension에 추가되어있는 fetch<ParsingType:Decodable>(url,completion:)메서드 구현부 수정

Hosinging committed:
chore: 오타 수정

Hosinging commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -30,14 +31,33 @@ enum APIError: LocalizedError {
 }
 
+enum URI {
+    static let host = "https://api.openweathermap.org/"
```
url에 대해 찾아봤을 때 아래 링크와 같은 내용들을 봤었습니다. 

[링크](https://www.betterweb.or.kr/blog/url%EC%9D%B4%EB%9E%80/)
[링크2](https://futurists.tistory.com/95)

그리고, URLComponents에서 scheme과 param사이에 들어가는 내용들을 host라는 프로퍼티로 사용하기도 해서 host라는 단어가 이상하지 않다고 생각했었던 것 같습니다. 

해당 내용들은 바로 아래 남겨주신 리뷰내용을 반영하면서 자연스럽게 URLComponents를 사용하게되서 URLComponents에서 사용하는 프로퍼티와 네이밍을 맞춰 분리해 주었습니다.

jaemuYeo commented on WeatherForecast/WeatherForecast/Data/WeatherDataManger.swift:
```diff
@@ -30,14 +31,33 @@ enum APIError: LocalizedError {
-    private let appId = "af361cc4ac7bf412119174d64ba296ff"
-    
     private init() {}
+    private var isPathCurrent = true
```
위에 리뷰말씀처럼 파라미터들을 분리하게 되면서 Path를 구분하는 flag도
지금 리뷰말씀처럼 변경하였습니다!

Hosinging committed:
chore: 잘못 삭제한 메서드 복구

Hosinging committed:
feat: 상수 선언 및 싱글턴 객체 바인딩

Hosinging committed:
chore: 파일 내 생성 문구 수정

Hosinging committed:
chore: 불필요 코드 제거

wonhee009 commented on WeatherForecast/WeatherForecast/Controllers/WeatherListViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        switch status {
+        case .authorizedWhenInUse, .authorizedAlways:
+            currentLocation = manager.location?.coordinate
+            guard let currentLocation = currentLocation else { return }
+            weatherDataManater.longitude = currentLocation.longitude
+            weatherDataManater.latitude = currentLocation.latitude
```
`currentLocation` 프로퍼티가 해당 영역에서만 사용되는거 같은데 전역변수로 선언한 이유가 있을까요?

wonhee009 commented on WeatherForecast/WeatherForecast/Controllers/WeatherListViewController.swift:
```diff
@@ -0,0 +1,91 @@
+class WeatherListViewController: UIViewController {
+    private var locationManager: CLLocationManager?
+    private var currentLocation: CLLocationCoordinate2D?
+    private var weatherDataManater = WeatherDataManager.shared
```
`shared`라고 네이밍한거로 보아 싱글톤 패턴으로 정의한걸로 보입니다.
`weatherDataManager`라는 프로퍼티 없이 `WeatherDataManager.shared.latitude` 혹은 `WeatherDataManager.shared.longitude`를 사용해도 동일하게 동작할텐데요.

싱글톤이란 무엇인지 그리고 프로퍼티로 한번 더 선언할 필요가 있었을지 한 번 생각해보시면 좋을거 같습니다.😀

wonhee009 replied:
그리고 프로퍼티 네이밍에 오타가 있네요😀

wonhee009 commented on WeatherForecast/WeatherForecast/Data/URLBuilder.swift:
```diff
@@ -0,0 +1,47 @@
+import Foundation
+
+class URLBuilder: URLMakable {
+    var pathType: URLResource.PathType = .current
```
`pathType`이 기본적으로 current로 정의되어져 있는데요.
5일치 예보를 보기 위해서는 URLBuilder 인스턴스의 pathType을 변경해줘야 하는건가요?

```swift
class URLBuilder: URLMakable {
     var pathType: URLResource.PathType

     init(pathType: URLResource.PathType) {
          self.pathType = pathType
     }
}
```

이런 식으로 init 시점에 지정해주는건 어떨까요?

wonhee009 commented on WeatherForecast/WeatherForecast/Data/URLBuilder.swift:
```diff
@@ -0,0 +1,47 @@
+        urlComponents.scheme = resource.scheme
+        urlComponents.host = resource.host
+        
+        switch pathType {
+        case .current:
+            urlComponents.path = URLResource.PathType.current.rawValue
+        case .fiveDays:
+            urlComponents.path = URLResource.PathType.fiveDays.rawValue
+        }
```
```swift
urlComponents.path = pathType.rawValue
```
로 가능할거 같습니다.
